KIPs/kip-114.md

```diff
@@ -0,0 +1,324 @@
+        assert verified
+        print("verified:", verified)
+        print("=" * 80)
+        prevHeader = header
+
+
+main()
+```
+
+### Genesis Block
+
+If `FORK_BLOCK` is zero, the genesis block must have `randomReveal`, `mixHash` fields.
+
+## Rationale
+
+### Names of the header fields
+
+The naming is highly inspired by [EIP-4339](https://eips.ethereum.org/EIPS/eip-4399).
+We use `randomReveal` instead of RANDAO reveal because the randoms are generated in a different manner.
+
+### Block size increment due to new fields in the header
+
+The size increment cost of RLP-encoded header is around 131 bytes per block, which is approximately 3.8GB per year. It is a negligible cost compared to its benefit.
+
+### BLS-based random over EC-based random
+
+We chose BLS-based verifiable random over EC-based because:
+
+- A BLS library written in Golang is easy to find as in [link](https://github.com/klaytn/klaytn/issues/1823), whereas it is not the case for EC-based as in [link](https://www.ietf.org/archive/id/draft-irtf-cfrg-vrf-15.html#section-6).
+- There is a potential consensus change based on BLS signature. Therefore, this proposal gently introduces BLS signature to the Klaytn protocol as well as adding on-chain randomness.
+
+## Backwards Compatibility
+
+Blocks before `FORK_BLOCK` will remain the same.
+
+## Security Considerations
+
+### Biasability
+
+BLS signature as well as the signing message (i.e., proposing block number) is deterministic. Thus, the proposer has no influence over the computation of `mixHash`. In addition, the biasability of `mixHash` mainly depends on that of the hash of `randomReveal`. To the best of our knowledge, keccak256 is not a biased hash function. Therefore, `mixHash`, which is the XOR of the output of keccak256 and the previous `mixHash`, is unbiased.
+
+### Predictability
+
+`mixHash` of the next block can be computed when the both requirements are met: (1) the `mixHash` of the latest block is revealed; and (2) the `randomReveal` of the next block is revealed. Even if the proposer of the next block can precompute the latter as soon as the proposer found out that it is proposing the next block, it does not know the former which circumvents the proposer from predicting `mixHash` of the next block. In other words, only after the proposer receives the latest `mixHash`, it is able to know the `mixHash` of the next block.
+
+However, in reality, there is a possibility of collusion among validators that facilitates the prediction beyond the ideal situation. Specifically, colluding validators can share their `randomReveal`s in advance to enable the precomputation of `mixHash`. The extent to which colluding validators can predict depends on the number of consecutive blocks they collectively propose; the more the validators collude, the further they can predict. A single benign validator can disrupt the collusion, thereby limiting the predictable timeframe.
```

Comment:
Is there any way of detecting colluding happens when we investigate the blocks?

Reply:
I don't think it's possible. Looking at those values will not tell us anything because they are deterministic. We can suspect the collusion if the validators always succeeds in front-running, but it's only a conjecture.

Reply:
okay. I see. Thanks.